### ISSUE TITLE
Current implementation of recursive chmod considered harmful

### ISSUE THREAD

**Issue by allisonkarlitskaya** [cockpit-project/cockpit-files#192]
We currently have a button for "Change Permissions for Enclosed Files" which calls `chmod -R` with a numeric mode.  That's a problem because you usually don't want files and directories to have the same mode.  Either all your files end up executable or all of your directories end up broken.

In GNOME this button is labelled "Change Permissions for Enclosed Files..." and opens a second dialog:

![image](https://github.com/cockpit-project/cockpit-navigator/assets/36541154/1e7fb20b-1ecf-4210-badd-ff45db339187)

If we're going to have recursive mode changing at all, we need something like this.  Until that point, I'd argue that we should remove the existing button.

**Comment by garrett** [comment]
I agree.

**Comment by allisonkarlitskaya** [comment]
`chmod` also has a way to say `+X` which means "add executable permission if any other user has executable permissions".  That usually does the "right thing" for both files and directories (and also includes files that are executable files).

If we dumb down the dialog a bit and make it about read/write permissions only, and use +X in the same places we allow read permissions, it might make even more sense than having the full-blown GNOME approach (which breaks existing executable files).

**Comment by martinpitt** [comment]
Agreed. I also added that to the MVP milestone.

**Comment by allisonkarlitskaya** [comment]
So here's a very concrete proposal for what I think should happen:

Files and directories should be treated exactly the same, always.  To state it explicitly: we should forget about the idea of having "list" (`+r`) and "access" (`+x`) treated separately for directories.  That's just "read", as far as we're concerned.

In any UI to deal with this, we'll have three categories: user, group, other.  Each of those will have a dropdown containing the options "Read and write", "Read-only", "No access":

```
File permissions

   User     [  Read and write  | v ]

   Group    [  Read-only       | v ]

   Other    [  No access       | v ]

```

with each of the dropdowns looking like:

```
   User     [  Read and write      ]
            |  Read-only           |
            |  No access           |
```

From this dialog, we construct a string via concatenation, using the following approach:

```
  "User" = "u"
    "Read and write" = "+rwX"
    "Read-only" = "+rX-w"
    "No access" = "-rwx"
","
  "Group" = "g"
    ... as for the options for user ...
","
  "Other" = "o"
    ... as for the options for user ...
```

So given the example above, the resulting string would be `u+rwX,g+rX-w,o-rwx`.  This is suitable for passing to `chmod` or `chmod -R` for any file or directory.

**Comment by allisonkarlitskaya** [comment]
>  a checkbox in an inconsistent state (does Patternfly have that?)

https://www.patternfly.org/components/forms/checkbox/

> isChecked
> boolean | null
> false
> Flag to show if the checkbox is checked. If null, the checkbox will be indeterminate (partially checked).

**Comment by jelly** [comment]
> > a checkbox in an inconsistent state (does Patternfly have that?)
> 
> https://www.patternfly.org/components/forms/checkbox/
> 
> > isChecked
> > boolean | null
> > false
> > Flag to show if the checkbox is checked. If null, the checkbox will be indeterminate (partially checked).

No, use a Switch https://www.patternfly.org/components/switch/#checked-with-label :)

**Comment by martinpitt** [comment]
@jelly says that this functionality doesn't exist, so it doesn't need to block the first release.

**Comment by garrett** [comment]
This is similar to the mockups I already made in PF style and shared a while back. Here's the latest version:

![image](https://github.com/user-attachments/assets/034b81b0-dc73-40c8-8294-464319c53bcc)

BTW: It's a checkbox instead of a switch as switches are instant-apply and checkboxes are queued. See https://www.patternfly.org/components/switch/design-guidelines#when-to-use-switches-versus-checkboxes for details.

**Comment by garrett** [comment]
@jelly: Sure, that all sounds reasonable. Show it when the filetype can be executable (we should use the same logic), then use +X.

(I thought we talked about this in a call a few months ago and agreed on it, but it's good to have it clearly agreed upon and written down here regardless. Thanks for asking and formalizing it!)

I'm guessing binary files and shell scripts can be executable, as well as things like AppImages (which are mainly outside the scope of Cockpit, except if an admin is browsing a home dir, I suppose). I suppose it uses magic sniffing for mimetypes with an accept list.

**Comment by tomasmatus** [comment]
fixed in #775
